apps/rush-lib/src/logic/Git.ts

```diff
@@ -3,76 +3,100 @@
 
 import gitInfo = require('git-repo-info');
 import * as child_process from 'child_process';
+import * as os from 'os';
 
-import { PublishUtilities } from './PublishUtilities';
+import { Utilities } from '../utilities/Utilities';
+import { AlreadyReportedError } from '../utilities/AlreadyReportedError';
+import { GitEmailPolicy } from './policy/GitEmailPolicy';
+import { RushConfiguration } from '../api/RushConfiguration';
 
-export class Git {
-  private _targetBranch: string | undefined;
-
-  public static detectIfGitIsSupported(): boolean {
-    const command: string = process.platform === 'win32' ? 'where' : 'which';
-    const result: child_process.SpawnSyncReturns<string> = child_process.spawnSync(command, ['git']);
-
-    if (result.status !== 0) {
-      return false;
-    }
+interface IResultOrError<TResult> {
+  error?: Error;
+  result?: TResult;
+}
 
-    try {
-      return !!gitInfo().sha;
-    } catch (e) {
-      return false; // Unexpected, but possible if the .git directory is corrupted.
+export class Git {
+  private static _hasGit: boolean | undefined = undefined;
+  private static _gitPath: string | undefined;
+
+  public static getGitPath(): string | undefined {
+    if (Git._hasGit === undefined) {
+      const command: string = process.platform === 'win32' ? 'where' : 'which';
+      const result: child_process.SpawnSyncReturns<string> = child_process.spawnSync(command, ['git']);
+
+      if (result.status === 0) {
+        Git._gitPath = result.stdout;
+        Git._hasGit = !!result.stdout;
+      }
     }
-  }
-
-  constructor(targetBranch: string | undefined) {
-    this._targetBranch = targetBranch;
-  }
-
-  public checkout(branchName: string | undefined, createBranch?: boolean): void {
-    const params: string = `checkout ${createBranch ? '-b ' : ''}${branchName}`;
 
-    PublishUtilities.execCommand(!!this._targetBranch, 'git', params.split(' '));
+    return Git._gitPath;
   }
 
-  public merge(branchName: string): void {
-    PublishUtilities.execCommand(!!this._targetBranch, 'git', `merge ${branchName} --no-edit`.split(' '));
-  }
-
-  public deleteBranch(branchName: string, hasRemote: boolean = true): void {
-    PublishUtilities.execCommand(!!this._targetBranch, 'git', `branch -d ${branchName}`.split(' '));
-    if (hasRemote) {
-      PublishUtilities.execCommand(!!this._targetBranch, 'git', `push origin --delete ${branchName}`.split(' '));
+  public static detectIfGitIsSupported(): boolean {
```

Comment:
>detectIfGitIsSupported [](start = 16, length = 22)

Document that this checks whether the current working directory is part of a Git working folder.

Reply:
isInGitWorkingDirectory)?

---
In reply to: [213455262](https://github.com/Microsoft/web-build-tools/pull/786#discussion_r213455262) [](ancestors = 213455262)